### PR TITLE
Custom no-results message support

### PR DESCRIPTION
**Pull request by santhoshtr**
* Refactoring and clean up for LanguageCategoryDisplay class
* Document the options for LanguageCategoryDisplay class
* Reduce the spreading of no results handler code
* Add an option to accept no results template
* Remove unwanted, unused constructor too
* Use CSS to hide or show the no-results view
* Remove the unwanted noresults method in jquery.uls.core, directly
  call the same method of lcd.
* Add an example

We might need some context also to the custom message such as the current query string. But we can revisit it when we really design some custom message for some use cases in mediawiki.

### CODE REVIEW

File: src/jquery.uls.lcd.js [hunk 1]
```diff
@@ -363,22 +362,20 @@
 		},
 
 		noResults: function () {
-			var $suggestions;
-			this.$noResults.removeClass( 'hide' );
-			this.$noResults.siblings( '.uls-lcd-region-section' ).addClass( 'hide' );
+			var $suggestions, $noResults;
 
-			// Only build the data once
```
Comment (santhoshtr) [inline src/jquery.uls.lcd.js]: I had no idea why this check was done. The quickList is cached anyway. Also this was causing the issue: When I search for something and get no results everything fine, then I do a search and get some result. Then search for some non-existing language-The no-results screen then won't have suggestions.

I just removed this check

File: css/jquery.uls.lcd.css [hunk 2]
```diff
@@ -102,6 +102,18 @@
 	vertical-align: middle;
 }
 
+.uls-no-results-view {
+	display: none;
+}
+
+.uls-lcd.uls-no-results > .uls-lcd-region-section {
```
Comment (Nikerabbit) [inline css/jquery.uls.lcd.css]: Why is this needed? Shouldn't the layout of the no-results view be fully under control of the code generating it?

File: examples/test-no-results.html [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+		<!-- <link rel="apple-touch-icon" href="/apple-touch-icon.png"> -->
+		<!-- <meta name="description" content=""> -->
+		<meta name="author" content="Santhosh Thottingal">
+		<link href="../css/jquery.uls.css" rel="stylesheet">
+		<link href="../css/jquery.uls.grid.css" rel="stylesheet">
+		<link href="../css/jquery.uls.lcd.css" rel="stylesheet">
+		<!-- demo -->
+		<link href="resources/demo.css" rel="stylesheet">
+		<!-- Libs -->
+		<script src="https://code.jquery.com/jquery-1.11.1.min.js"></script>
```
Comment (Nikerabbit) [inline examples/test-no-results.html]: Old jquery?

File: examples/test-no-results.html [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+		<!-- Source -->
+		<script src="../src/jquery.uls.data.js"></script>
+		<script src="../src/jquery.uls.data.utils.js"></script>
+		<script src="../src/jquery.uls.lcd.js"></script>
+		<script src="../src/jquery.uls.languagefilter.js"></script>
+		<script src="../src/jquery.uls.core.js"></script>
+		<script>
+			$( document ).ready( function() {
+				$( '.uls-trigger' ).uls( {
+					onSelect : function( language ) {
```
Comment (Nikerabbit) [inline examples/test-no-results.html]: Please remove extra space before colon.

File: examples/test-no-results.html [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+						var languageName = $.uls.data.getAutonym( language );
+						$( '.uls-trigger' ).text( languageName );
+					},
+					noResultsTemplate: '<div>No article exist in the language you searched</div>'
```
Comment (Nikerabbit) [inline examples/test-no-results.html]: typo: exists

File: examples/test-no-results.html [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+				} );
+			} );
+		</script>
```
Comment (Nikerabbit) [inline examples/test-no-results.html]: This demo doesn't really illustrate the difference between "unknown language" and "no search results".

File: examples/test-no-results.html [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+	</head>
+
+	<body>
+		<div class="navbar navbar-fixed-top">
+			<span class="active uls-trigger">Select Language</span>
```
Comment (Nikerabbit) [inline examples/test-no-results.html]: I am not fan of title case here.

File: examples/test-no-results.html [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+			<h1>Universal Language Selector</h1>
+			<p>
+				Demonstration of jQuery plugin
+			</p>
+		</div>
+		<div class="container"></div>
```
Comment (Nikerabbit) [inline examples/test-no-results.html]: Unused?

File: src/jquery.uls.lcd.js [hunk 9]
```diff
@@ -22,43 +22,40 @@
...
+	/**
+	 * Language category display
+	 * @param {Element} element The container element to which the languages to be displayed
+	 * @param {Object} [options] Configuration object
+	 * @cfg {Object} [languages] Languages known to the ULS. Keyed by language code, values are autonyms.
+	 * @cfg {string[]} [showRegions] Array of region codes to show. Default is
+	 *  [ 'WW', 'AM', 'EU', 'ME', 'AF', 'AS', 'PA' ],
+	 * @cfg {number} [itemsPerColumn] Number of languages per column.
+	 * @cfg {number} [columns] Number of columns for languages. Default is 4.
+	 * @cfg {function} [languageDecorator] Callback function to be called when a language
```
Comment (Nikerabbit) [inline src/jquery.uls.lcd.js]: Functions are objects too, right? Should be Function. (also below)

File: src/jquery.uls.lcd.js [hunk 10]
```diff
@@ -22,43 +22,40 @@
...
+	 *  link is prepared - for custom decoration.
+	 * @cfg {function|string[]} [quickList] The languages to display as suggestions for quick selectoin.
+	 * @cfg {string|jQuery} [noResultsTemplate]
```
Comment (Nikerabbit) [inline src/jquery.uls.lcd.js]: Let's not accept strings. It's a bad practice, as there isn't sufficient escaping. I predict we will need to accept functions here so that we can pass parameters (for example to differentiate between no matches or unknown language).

File: src/jquery.uls.lcd.js [hunk 11]
```diff
@@ -363,22 +362,20 @@
...
-			$suggestions.find( 'h3' )
+			$suggestions
+				.removeClass( 'hide' )
+				.find( 'h3' )
 				.data( 'i18n', 'uls-no-results-suggestion-title' )
 				.text( 'You may be interested in:' )
 				.i18n();
-			this.$noResults.find( 'h2' ).after( $suggestions );
+
+			$noResults.find( 'h2' ).after( $suggestions );
```
Comment (Nikerabbit) [inline src/jquery.uls.lcd.js]: This looks very unreliable now that $noResults is completely out of our control.

File: src/jquery.uls.lcd.js [hunk 12]
```diff
@@ -363,22 +362,20 @@
...
-			this.$noResults.removeClass( 'hide' );
-			this.$noResults.siblings( '.uls-lcd-region-section' ).addClass( 'hide' );
+			var $suggestions, $noResults;
 
-			// Only build the data once
-			if ( this.$noResults.find( '.uls-lcd-region-title' ).length ) {
-				return;
-			}
+			this.$element.addClass( 'uls-no-results' );
+			$noResults = this.$element.children( '.uls-no-results-view' );
```
Comment (Nikerabbit) [inline src/jquery.uls.lcd.js]: Why children? why do we need DOM access at all?

File: examples/test-no-results.html (hunk 4 above)
Comment (santhoshtr) [reply examples/test-no-results.html]: :heavy_check_mark:

File: examples/test-no-results.html (hunk 5 above)
Comment (santhoshtr) [reply examples/test-no-results.html]: :heavy_check_mark:

File: examples/test-no-results.html (hunk 7 above)
Comment (santhoshtr) [reply examples/test-no-results.html]: :heavy_check_mark:

File: examples/test-no-results.html (hunk 8 above)
Comment (santhoshtr) [reply examples/test-no-results.html]: :heavy_check_mark:

File: src/jquery.uls.lcd.js (hunk 9 above)
Comment (santhoshtr) [reply src/jquery.uls.lcd.js]: :heavy_check_mark:

File: examples/test-no-results.html (hunk 3 above)
Comment (santhoshtr) [reply examples/test-no-results.html]: :heavy_check_mark:

File: css/jquery.uls.lcd.css (hunk 2 above)
Comment (santhoshtr) [reply css/jquery.uls.lcd.css]: I did not like the hide class usage, I just removed it and added a marker class to the lcd as .uls-no-results - that take care of hiding and showing of regiions. We still use it for regions, but I would like to remove the usage of hide class altogether. It is just a wrapper around display:none

File: examples/test-no-results.html (hunk 6 above)
Comment (santhoshtr) [reply examples/test-no-results.html]: Updated the demo to use the current search query

File: src/jquery.uls.lcd.js (hunk 10 above)
Comment (santhoshtr) [reply src/jquery.uls.lcd.js]: I improved the code to accept functions and passes the current search query as param

File: src/jquery.uls.lcd.js (hunk 11 above)
Comment (santhoshtr) [reply src/jquery.uls.lcd.js]: I moved this to the default handler for no-results

File: css/jquery.uls.lcd.css (hunk 2 above)
Comment (Nikerabbit) [reply css/jquery.uls.lcd.css]: Removing `hide` is good. I initially misread that this to mean we were hiding section headers in the no results template, but I see now this just hides everything else except the no results template.

File: src/jquery.uls.lcd.js [hunk 13]
```diff
@@ -362,23 +363,29 @@
...
+			this.$element.addClass( 'uls-no-results' );
+
+			this.$element.find( '.uls-no-results-view' ).remove();
 
-			// Only build the data once
-			if ( this.$noResults.find( '.uls-lcd-region-title' ).length ) {
-				return;
+			if ( typeof this.options.noResultsTemplate === 'function' ) {
+				$noResults =
+						this.options.noResultsTemplate.call( this, currentSearchQuery );
```
Comment (Nikerabbit) [inline src/jquery.uls.lcd.js]: There seems to be one extra tab on ths line.

File: src/jquery.uls.lcd.js [hunk 14]
```diff
@@ -22,43 +22,41 @@
...
+	 * @param {Object} [options] Configuration object
+	 * @cfg {Object} [languages] Languages known to the ULS. Keyed by language code, values are autonyms.
+	 * @cfg {string[]} [showRegions] Array of region codes to show. Default is
+	 *  [ 'WW', 'AM', 'EU', 'ME', 'AF', 'AS', 'PA' ],
+	 * @cfg {number} [itemsPerColumn] Number of languages per column.
+	 * @cfg {number} [columns] Number of columns for languages. Default is 4.
+	 * @cfg {Function} [languageDecorator] Callback function to be called when a language
+	 *  link is prepared - for custom decoration.
+	 * @cfg {Function|string[]} [quickList] The languages to display as suggestions for quick selectoin.
+	 * @cfg {jQuery} [noResultsTemplate]
```
Comment (Nikerabbit) [inline src/jquery.uls.lcd.js]: {jQuery|Function}
